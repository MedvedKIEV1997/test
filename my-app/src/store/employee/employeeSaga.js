import { all, call, put, select, takeLatest } from 'redux-saga/effects';
import axios from 'axios';
import {
    getEmployeeFailure,
    getEmployeeFetch,
    getEmployeeSuccess,
    getSearchFailure,
    getSearchFetch,
    getSearchSuccess,
    getSubordinatesFailure,
    getSubordinatesFetch,
    getSubordinatesSuccess,
    selectSubordinates
} from './employeeSlice';

function* workGetSearchFetch({ payload }) {
    try {
        const response = yield call(
            axios.get,
            `http://localhost:8000/employees/search/${payload}`
        );
        yield put(getSearchSuccess(response.data));
    } catch (error) {
        yield put(getSearchFailure(error));
    }
}

function* workGetEmployeeFetch({ payload }) {
    try {
        const response = yield call(
            axios.get,
            `http://localhost:8000/employees/${payload}`
        );
        yield put(getEmployeeSuccess(response.data));
        yield call(workGetSubordinatesFetch);
    } catch (error) {
        yield put(getEmployeeFailure(error));
    }
}

function* workGetSubordinatesFetch() {
    try {
        const subordinates = yield select(selectSubordinates);
        if (subordinates.length !== 0) {
            const response = yield all(
                subordinates.map((id) => {
                    const response = call(
                        axios.get,
                        `http://localhost:8000/employees/${id}`
                    );
                    return response;
                })
            );

            const subordinatesData = response.map((el) => el.data);
            console.log(subordinatesData);
            yield put(getSubordinatesSuccess(subordinatesData));
            yield call(workGetSubordinatesFetch);
        }
    } catch (error) {
        yield put(getSubordinatesFailure(error));
    }
}

function* onSearchFetch() {
    yield takeLatest(getSearchFetch.type, workGetSearchFetch);
}

function* onEmployeeFetch() {
    yield takeLatest(getEmployeeFetch.type, workGetEmployeeFetch);
}

function* onSubordinatesFetch() {
    yield takeLatest(getSubordinatesFetch.type, workGetSubordinatesFetch);
}

export function* employeeSaga() {
    yield all([
        call(onEmployeeFetch),
        call(onSubordinatesFetch),
        call(onSearchFetch)
    ]);
}

export default employeeSaga;
