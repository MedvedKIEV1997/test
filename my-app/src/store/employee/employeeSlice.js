import { createSlice } from '@reduxjs/toolkit';

const employeeSlice = createSlice({
  name: 'employee',
  initialState: {
    employee: null,
    subordinates: [],
    subordinatesToShow: [],

    error: null,
  },
  reducers: {
    getEmployeeFetch: (state, action) => {
      state.employee = null;
      state.subordinates = [];
      state.error = null;
    },
    getEmployeeSuccess: (state, action) => {
      state.employee = action.payload;
      state.subordinates = action.payload.controls;
      state.subordinatesToShow = action.payload.controls;
    },
    getEmployeeFailure: (state, action) => {
      state.error = action.payload;
      state.subordinates = [];
      state.subordinatesToShow = [];
    },
    getSubordinatesFetch: (state, action) => {
      state.subordinatesToShow = state.subordinates;
    },
    getSubordinatesSuccess: (state, action) => {
      state.subordinatesToShow = [
        ...state.subordinatesToShow,
        ...action.payload,
      ];
      state.subordinates = action.payload;
    },
    getSubordinatesFailure: (state, action) => {
      state.error = action.payload;
      state.subordinates = [];
      state.subordinatesToShow = [];
    },
  },
});

export const {
  getEmployeeFailure,
  getEmployeeSuccess,
  getEmployeeFetch,
  getSubordinatesFailure,
  getSubordinatesFetch,
  getSubordinatesSuccess,
} = employeeSlice.actions;

export const selectEmployee = (state) => state.employee.employee;
export const selectSubordinates = (state) => state.employee.subordinates;
export const selectSubordinatesToShow = (state) =>
  state.employee.subordinatesToShow;

export default employeeSlice.reducer;
