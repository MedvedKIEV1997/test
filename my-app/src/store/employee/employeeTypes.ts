export type Employee = {
    name: string;
    controls: number[];
    id: number;
};

export type SearchEmployeesPayload = {
    name: string;
};

export type GetEmployeePayload = {
    id: number;
};
