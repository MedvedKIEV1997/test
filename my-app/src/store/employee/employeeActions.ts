import { createAction } from '@reduxjs/toolkit';
import { GetEmployeePayload, SearchEmployeesPayload } from './employeeTypes';

export const searchEmployees =
    createAction<SearchEmployeesPayload>('searchEmployees');

export const getEmployee = createAction<GetEmployeePayload>('getEmployee');

export const getSubordinates = createAction('getSubordinates');
