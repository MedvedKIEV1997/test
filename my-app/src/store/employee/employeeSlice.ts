import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';
import { Employee } from './employeeTypes';

type InitialState = {
    employee: Employee | null;
    subordinates: number[];
    subordinatesToShow: Employee[];
    search: Employee[];
};

const initialState: InitialState = {
    employee: null,
    subordinates: [],
    subordinatesToShow: [],
    search: []
};

const employeeSlice = createSlice({
    name: 'employee',
    initialState,
    reducers: {
        setSearch: (state, { payload }: PayloadAction<Employee[]>) => {
            state.search = payload;
        },

        setEmployee: (state, { payload }: PayloadAction<Employee>) => {
            state.subordinatesToShow = [];
            state.employee = payload;
            state.subordinates = payload.controls;
        },
        setSubordinates: (state, { payload }: PayloadAction<Employee[]>) => {
            state.subordinatesToShow = [
                ...state.subordinatesToShow,
                ...payload
            ];
            state.subordinates = payload.map((el) => el.controls).flat();
        }
    }
});

export const sliceActions = employeeSlice.actions;

export const selectEmployee = (state: RootState) => state.employee.employee;
export const selectSubordinates = (state: RootState) =>
    state.employee.subordinates;
export const selectSubordinatesToShow = (state: RootState) =>
    state.employee.subordinatesToShow;
export const selectSearch = (state: RootState) => state.employee.search;

export default employeeSlice.reducer;
