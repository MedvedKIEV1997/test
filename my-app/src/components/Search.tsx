import { ChangeEvent, useEffect, useState } from 'react';
import { Link, Outlet } from 'react-router-dom';
import { Col, Form, Row, Stack, Container, Button } from 'react-bootstrap';

import useActions from '../hooks/useActions';
import useTypedSelector from '../hooks/useTypedSelectors';
import { Employee, selectEmployees } from '../redux/ducks/employee.duck';
import { CustomButton } from '../styled/search.styles';

const Search = () => {
    const [employeesToShow, setEmployeesToShow] = useState<Employee[]>([]);

    const allEmployees = useTypedSelector(selectEmployees);
    const actions = useActions();

    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
        setEmployeesToShow(
            allEmployees.filter((employee) =>
                employee.name
                    .toLocaleLowerCase()
                    .includes(event.target.value.toLocaleLowerCase())
            )
        );
    };

    useEffect(() => {
        actions.getEmployees();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    useEffect(() => {
        setEmployeesToShow(allEmployees);
    }, [allEmployees]);

    return (
        <Container className="p-5" fluid>
            <Row>
                <Col xs={12} md={5} lg={4} xl={3}>
                    <Stack gap={3}>
                        <Form.Control
                            placeholder="Write name here..."
                            onChange={handleChange}
                        />

                        {employeesToShow.length !== 0 && (
                            <Row className="g-2 ">
                                {employeesToShow.map((employee) => (
                                    <Col
                                        xl={3}
                                        md={4}
                                        sm={2}
                                        xs={4}
                                        key={employee.id}
                                        className="d-flex"
                                    >
                                        <Link
                                            to={'' + employee.id}
                                            className="w-100"
                                        >
                                            <CustomButton variant="outline-primary">
                                                {employee.name}
                                            </CustomButton>
                                        </Link>
                                    </Col>
                                ))}
                            </Row>
                        )}
                    </Stack>
                </Col>
                <Col>
                    <Outlet />
                </Col>
            </Row>
        </Container>
    );
};

export default Search;
