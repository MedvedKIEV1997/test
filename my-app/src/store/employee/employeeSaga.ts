import { all, call, put, select, takeLatest } from 'redux-saga/effects';
import axios, { AxiosResponse } from 'axios';

import * as actions from './employeeActions';
import { selectSubordinates, sliceActions } from './employeeSlice';
import { Employee } from './employeeTypes';

function* searchEmployees(action: ReturnType<typeof actions.searchEmployees>) {
    const {
        payload: { name }
    } = action;
    try {
        const { data }: AxiosResponse<Employee[]> = yield call(
            axios.get,
            `http://localhost:8000/employees/search/${name}`
        );
        yield put(sliceActions.setSearch(data));
    } catch (error) {
        console.log(error);
    }
}

function* getEmployee(action: ReturnType<typeof actions.getEmployee>) {
    const {
        payload: { id }
    } = action;

    try {
        const { data }: AxiosResponse<Employee> = yield call(
            axios.get,
            `http://localhost:8000/employees/${id}`
        );

        yield put(sliceActions.setEmployee(data));

        yield call(getSubordinates);
    } catch (error) {
        console.log(error);
    }
}

//I've tried to get the type
function* getSubordinates(): any {
    try {
        const subordinates: ReturnType<typeof selectSubordinates> =
            yield select(selectSubordinates);

        if (subordinates.length !== 0) {
            const response: AxiosResponse[] = yield all(
                subordinates.map((id) => {
                    const response = call(
                        axios.get,
                        `http://localhost:8000/employees/${id}`
                    );
                    return response;
                })
            );

            const subordinatesData: Employee[] = response.map((el) => el.data);

            yield put(sliceActions.setSubordinates(subordinatesData));

            yield* getSubordinates();
        }
    } catch (error) {
        console.log(error);
    }
}

function* watchSearch() {
    yield takeLatest(actions.searchEmployees.type, searchEmployees);
}

function* watchGetEmployee() {
    yield takeLatest(actions.getEmployee.type, getEmployee);
}

function* employeeSaga() {
    yield all([call(watchGetEmployee), call(watchSearch)]);
}

export default employeeSaga;
