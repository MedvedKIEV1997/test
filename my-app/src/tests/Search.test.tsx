import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Search from '../components/Search';
import { employees, renderWithProviders } from './test-utils';
import { initialState } from '../redux/ducks/employee.duck';

describe('search component tests', () => {
    const setup = () => {
        return {
            user: userEvent.setup(),
            ...renderWithProviders(<Search />, {
                preloadedState: {
                    employee: {
                        ...initialState,
                        allEmployees: employees
                    }
                }
            })
        };
    };

    it('should match a snapshot', () => {
        const { asFragment } = setup();
        expect(asFragment()).toMatchSnapshot();
    });

    it('should filter all employees and get Link "Lin Cheng"', async () => {
        const { user } = setup();
        const searchBar = screen.getByPlaceholderText(/Write name here/i);
        await user.type(searchBar, 'lin');
        expect(
            screen.getByRole('button', { name: 'Lin Cheng' })
        ).toBeInTheDocument();
    });

    it('should filter all employees and do not find Link "Dew Kelg"', async () => {
        const { user } = setup();
        const searchBar = screen.getByPlaceholderText(/Write name here/i);
        await user.type(searchBar, 'lin');
        expect(
            screen.queryByRole('button', { name: 'Dew Kelg' })
        ).not.toBeInTheDocument();
    });
});
