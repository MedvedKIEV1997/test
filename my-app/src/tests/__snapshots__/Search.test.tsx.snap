// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search component tests should match a snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-5 container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col-xl-3 col-lg-4 col-md-5 col-12"
      >
        <div
          class="vstack gap-3 minBreakpoint-xs"
        >
          <input
            class="form-control"
            placeholder="Write name here..."
          />
          <div
            class="g-2  row"
          >
            <div
              class="d-flex col-xl-3 col-md-4 col-sm-2 col-4"
            >
              <a
                class="w-100"
                href="/1"
              >
                <button
                  class="sc-bczRLJ gMmpAf btn btn-outline-primary"
                  type="button"
                >
                  Doug Lime
                </button>
              </a>
            </div>
            <div
              class="d-flex col-xl-3 col-md-4 col-sm-2 col-4"
            >
              <a
                class="w-100"
                href="/2"
              >
                <button
                  class="sc-bczRLJ gMmpAf btn btn-outline-primary"
                  type="button"
                >
                  Lin Cheng
                </button>
              </a>
            </div>
            <div
              class="d-flex col-xl-3 col-md-4 col-sm-2 col-4"
            >
              <a
                class="w-100"
                href="/3"
              >
                <button
                  class="sc-bczRLJ gMmpAf btn btn-outline-primary"
                  type="button"
                >
                  Bob Welp
                </button>
              </a>
            </div>
            <div
              class="d-flex col-xl-3 col-md-4 col-sm-2 col-4"
            >
              <a
                class="w-100"
                href="/4"
              >
                <button
                  class="sc-bczRLJ gMmpAf btn btn-outline-primary"
                  type="button"
                >
                  Dakota Stein
                </button>
              </a>
            </div>
            <div
              class="d-flex col-xl-3 col-md-4 col-sm-2 col-4"
            >
              <a
                class="w-100"
                href="/5"
              >
                <button
                  class="sc-bczRLJ gMmpAf btn btn-outline-primary"
                  type="button"
                >
                  Mongo K
                </button>
              </a>
            </div>
            <div
              class="d-flex col-xl-3 col-md-4 col-sm-2 col-4"
            >
              <a
                class="w-100"
                href="/6"
              >
                <button
                  class="sc-bczRLJ gMmpAf btn btn-outline-primary"
                  type="button"
                >
                  Dew Dong
                </button>
              </a>
            </div>
            <div
              class="d-flex col-xl-3 col-md-4 col-sm-2 col-4"
            >
              <a
                class="w-100"
                href="/7"
              >
                <button
                  class="sc-bczRLJ gMmpAf btn btn-outline-primary"
                  type="button"
                >
                  Mort Dog
                </button>
              </a>
            </div>
            <div
              class="d-flex col-xl-3 col-md-4 col-sm-2 col-4"
            >
              <a
                class="w-100"
                href="/8"
              >
                <button
                  class="sc-bczRLJ gMmpAf btn btn-outline-primary"
                  type="button"
                >
                  Forman Go
                </button>
              </a>
            </div>
            <div
              class="d-flex col-xl-3 col-md-4 col-sm-2 col-4"
            >
              <a
                class="w-100"
                href="/9"
              >
                <button
                  class="sc-bczRLJ gMmpAf btn btn-outline-primary"
                  type="button"
                >
                  Dew Kelg
                </button>
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col"
      />
    </div>
  </div>
</DocumentFragment>
`;
